File: src/csharp/Examples/CurvatureAnalysis.cs
// CurvatureAnalysis.cs
// Example: Computing and analyzing discrete Gaussian curvature
// Added by Graph Technologies, 2025

using System;
using System.Linq;
using System.Numerics;
using DDGCompanion.Core;
using DDGCompanion.Algorithms;

namespace DDGCompanion.Examples
{
    public class CurvatureAnalysis
    {
        public static void Run()
        {
            Console.WriteLine("DDG Example: Curvature Analysis");
            Console.WriteLine("================================\n");
            
            // Create cube mesh
            var positions = new Vector3[]
            {
                new(-1, -1, -1), new(1, -1, -1),
                new(1,  1, -1), new(-1,  1, -1),
                new(-1, -1,  1), new(1, -1,  1),
                new(1,  1,  1), new(-1,  1,  1)
            };
            
            var faces = new int[,]
            {
                { 0, 1, 2 }, { 0, 2, 3 },  // Front
                { 4, 7, 6 }, { 4, 6, 5 },  // Back
                { 0, 4, 5 }, { 0, 5, 1 },  // Bottom
                { 2, 6, 7 }, { 2, 7, 3 },  // Top
                { 0, 3, 7 }, { 0, 7, 4 },  // Left
                { 1, 5, 6 }, { 1, 6, 2 }   // Right
            };
            
            var mesh = new Mesh();
            mesh.Build(positions, faces);
            
            Console.WriteLine("Mesh: Cube");
            Console.WriteLine($"  V = {mesh.Vertices.Count}, E = {mesh.Edges.Count}, F = {mesh.Faces.Count}");
            Console.WriteLine($"  χ = {mesh.EulerCharacteristic()}\n");
            
            // Compute Gaussian curvature
            var K = DiscreteGaussianCurvature.Compute(mesh);
            
            Console.WriteLine("Gaussian Curvature per vertex:");
            for (int i = 0; i < K.Length; i++)
            {
                double degrees = K[i] * 180.0 / Math.PI;
                Console.WriteLine($"  v[{i}]: K = {K[i]:F6} ({degrees:F2}°)");
            }
            
            // Verify Gauss-Bonnet
            double totalK = K.Sum();
            double expectedK = 2.0 * Math.PI * mesh.EulerCharacteristic();
            double error = Math.Abs(totalK - expectedK);
            
            Console.WriteLine("\nGauss-Bonnet Verification:");
            Console.WriteLine($"  Total curvature: {totalK:F6}");
            Console.WriteLine($"  Expected (2πχ):   {expectedK:F6}");
            Console.WriteLine($"  Error: {error:F6}");
            
            if (error < 0.01)
            {
                Console.WriteLine("  ✓ Gauss-Bonnet theorem verified!");
            }
            else
            {
                Console.WriteLine("  ✗ Error exceeds tolerance");
            }
        }
    }
}

File: src/cpp/examples/curvature_analysis.cpp
// curvature_analysis.cpp
// Example: Computing and analyzing discrete Gaussian curvature
// Added by Graph Technologies, 2025

#include <iostream>
#include <Eigen/Dense>
#include "../core/Mesh.h"
#include "../algorithms/DiscreteGaussianCurvature.h"

using namespace ddg;

int main() {
    std::cout << "DDG Example: Curvature Analysis\n";
    std::cout << "================================\n\n";
    
    // Create cube mesh
    Eigen::MatrixXd V(8, 3);
    V << -1, -1, -1,
          1, -1, -1,
          1,  1, -1,
         -1,  1, -1,
         -1, -1,  1,
          1, -1,  1,
          1,  1,  1,
         -1,  1,  1;
    
    Eigen::MatrixXi F(12, 3);
    F << 0, 1, 2,  0, 2, 3,
         4, 7, 6,  4, 6, 5,
         0, 4, 5,  0, 5, 1,
         2, 6, 7,  2, 7, 3,
         0, 3, 7,  0, 7, 4,
         1, 5, 6,  1, 6, 2;
    
    Mesh mesh;
    mesh.build(V, F);
    
    std::cout << "Mesh: Cube" << std::endl;
    std::cout << "  V = " << mesh.numVertices() << ", ";
    std::cout << "E = " << mesh.numEdges() << ", ";
    std::cout << "F = " << mesh.numFaces() << std::endl;
    std::cout << "  χ = " << mesh.eulerCharacteristic() << std::endl << std::endl;
    
    // Compute Gaussian curvature
    Eigen::VectorXd K = DiscreteGaussianCurvature::compute(mesh);
    
    std::cout << "Gaussian Curvature per vertex:\n";
    for (int i = 0; i < K.size(); i++) {
        std::cout << "  v[" << i << "]: K = " << K(i) 
                  << " (" << (K(i) * 180.0 / M_PI) << "°)" << std::endl;
    }
    
    // Verify Gauss-Bonnet theorem
    double totalK = DiscreteGaussianCurvature::totalCurvature(mesh);
    double expectedK = 2.0 * M_PI * mesh.eulerCharacteristic();
    
    std::cout << "\nGauss-Bonnet Verification:\n";
    std::cout << "  Total curvature: " << totalK << std::endl;
    std::cout << "  Expected (2πχ):   " << expectedK << std::endl;
    std::cout << "  Error: " << std::abs(totalK - expectedK) << std::endl;
    
    if (std::abs(totalK - expectedK) < 0.01) {
        std::cout << "  ✓ Gauss-Bonnet theorem verified!" << std::endl;
    } else {
        std::cout << "  ✗ Gauss-Bonnet error exceeds tolerance" << std::endl;
    }
    
    return 0;
}
